pkg/services/notary/notary.go

```diff
@@ -0,0 +1,364 @@
+	// We stop trying to send mainTx to the network if the chain reaches minNotValidBefore height.
+	minNotValidBefore uint32
+	fallbacks         []*transaction.Transaction
+	// nSigs is the number of signatures to be collected.
+	// nSigs == nKeys for standard signature request;
+	// nSigs <= nKeys for multisignature request.
+	nSigs uint8
+	// nSigsCollected is the number of already collected signatures
+	nSigsCollected uint8
+
+	// sigs is a map of partial multisig invocation scripts [opcode.PUSHDATA1+64+signatureBytes] grouped by public keys
+	sigs map[*keys.PublicKey][]byte
+}
+
+// NewNotary returns new Notary module.
+func NewNotary(bc blockchainer.Blockchainer, log *zap.Logger) (*Notary, error) {
+	cfg := bc.GetConfig().P2PNotary
+	w := cfg.UnlockWallet
+	wallet, err := wallet.NewWalletFromFile(w.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	haveAccount := false
+	for _, acc := range wallet.Accounts {
+		if err := acc.Decrypt(w.Password); err == nil {
+			haveAccount = true
+			break
+		}
+	}
+	if !haveAccount {
+		return nil, errors.New("no wallet account could be unlocked")
+	}
+
+	return &Notary{
+		requests: make(map[util.Uint256]request),
+		Config: Config{
+			MainCfg: cfg,
+			Chain:   bc,
+			Log:     log,
+		},
+		wallet: wallet,
+	}, nil
+}
+
+// OnNewRequest is a callback method which is called after new notary request is added to the notary request pool.
+func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	var isValid bool
+	nvbFallback := payload.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	nKeys := payload.MainTransaction.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+	typ, nSigs, pubs, err := n.verifyIncompleteWitnesses(payload.MainTransaction, nKeys)
+	if err == nil { // valid main tx signatures
+		isValid = true
+	}
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	r, exists := n.requests[payload.MainTransaction.Hash()]
+	if exists {
+		if r.nSigsCollected >= nSigs { // already collected sufficient number of signatures to complete main transaction
+			return // TODO: however, we may still accept extra fallback transactions without adding new signatures to main tx
+		}
+		r.isValid = isValid && r.isValid
+		if nvbFallback < r.minNotValidBefore {
+			r.minNotValidBefore = nvbFallback
+		}
+	} else {
+		r = request{
+			nSigs:             nSigs,
+			main:              payload.MainTransaction,
+			typ:               typ,
+			isValid:           isValid,
+			minNotValidBefore: nvbFallback,
+		}
+	}
+	if r.isValid {
+	loop:
+		for i, w := range payload.MainTransaction.Scripts {
+			if len(w.InvocationScript) != 0 {
+				switch r.typ {
+				case Signature:
+					if !exists || len(r.main.Scripts[i].InvocationScript) == 0 { // need this check because signature can already be added (consider receiving the same payload multiple times)
```

Comment:
`if !exists` you're doing `main:              payload.MainTransaction`, so do we need to do anything at all here (given that it's sequential, being done under lock)?

Reply:
We need to count collected signatures. So I just split this check into two different checks.